ouroboros-consensus/src/Ouroboros/Consensus/Node/Recovery.hs

```diff
@@ -0,0 +1,141 @@
+import           Ouroboros.Storage.ImmutableDB.Types (ImmutableDBError)
+import qualified Ouroboros.Storage.ImmutableDB.Types as ImmDB
+import           Ouroboros.Storage.VolatileDB.Types (VolatileDBError)
+import qualified Ouroboros.Storage.VolatileDB.Types as VolDB
+
+-- | The path to the /clean shutdown marker file/.
+cleanShutdownMarkerFile :: FsPath
+cleanShutdownMarkerFile = mkFsPath ["clean"]
+
+-- | When the given action terminates with a /clean/ exception, create the
+-- /clean shutdown marker file/.
+--
+-- NOTE: we assume the action (i.e., the node itself) never terminates without
+-- an exception.
+--
+-- A /clean/ exception is an exception for 'exceptionRequiresRecovery' returns
+-- 'False'.
+createMarkerOnCleanShutdown
+  :: IOLike m
+  => HasFS m h
+  -> m a  -- ^ Action to run
+  -> m a
+createMarkerOnCleanShutdown mp = onExceptionIf
+    (not . exceptionRequiresRecovery)
+    (createCleanShutdownMarker mp)
+
+-- | Return 'True' when 'cleanShutdownMarkerFile' exists.
+hasCleanShutdownMarker
+  :: IOLike m
+  => HasFS m h
+  -> m Bool
+hasCleanShutdownMarker hasFS =
+    doesFileExist hasFS cleanShutdownMarkerFile
+
+-- | Create the 'cleanShutdownMarkerFile'.
+--
+-- Will throw an 'FsResourceAlreadyExist' error when it already exists.
+createCleanShutdownMarker
+  :: IOLike m
+  => HasFS m h
+  -> m ()
+createCleanShutdownMarker hasFS = do
+    alreadyExists <- hasCleanShutdownMarker hasFS
+    unless alreadyExists $
+      withFile hasFS cleanShutdownMarkerFile (WriteMode MustBeNew) $ \_h ->
+        return ()
+
+-- | Remove 'cleanShutdownMarkerFile'.
+--
+-- Will throw an 'FsResourceDoesNotExist' error when it does not exist.
+removeCleanShutdownMarker
+  :: IOLike m
+  => HasFS m h
+  -> m ()
+removeCleanShutdownMarker hasFS =
+    removeFile hasFS cleanShutdownMarkerFile
+
+-- | Return 'True' if the given exception indicates that recovery of the
+-- database is required on the next startup.
+exceptionRequiresRecovery :: SomeException -> Bool
+exceptionRequiresRecovery e
+    | Just (ExceptionInLinkedThread _ e') <- fromException e
+    = exceptionRequiresRecovery e'
+    | Just (e' :: ChainDbFailure) <- fromException e
+    = case e' of
+        ImmDbFailure ue -> immDbUnexpectedError ue
+        VolDbFailure ue -> volDbUnexpectedError ue
+        LgrDbFailure fe -> fsError fe
+        _               -> recoveryRequired
+
+    -- The three exceptions below will always be wrapped in a
+    -- 'ChainDbFailure', but we include them just in case.
+    | Just (e' :: ImmutableDBError) <- fromException e
+    = case e' of
+        ImmDB.UnexpectedError ue -> immDbUnexpectedError ue
+        _                        -> noRecoveryRequired
+    | Just (e' :: VolatileDBError) <- fromException e
+    = case e' of
+        VolDB.UnexpectedError ue -> volDbUnexpectedError ue
+        _                        -> noRecoveryRequired
+    | Just (e' :: FsError) <- fromException e
+    = fsError e'
+
+    | otherwise
+    = noRecoveryRequired
```

Comment:
It makes me slightly uneasy that the _default_ is "no recovery required". But I guess it makes sense.